src/EditorFeatures/Core/Tagging/AsynchronousViewportTaggerProvider.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Tagging;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+/// <summary>
+/// Base type for async taggers that perform their work based on what code is current visible in the user's
+/// viewport.  These taggers will compute tags for what is visible at their specified <see
+/// cref="AbstractAsynchronousTaggerProvider{TTag}.EventChangeDelay"/>, but will also compute tags for the regions
+/// above and below that visible section at a delay of <see cref="DelayTimeSpan.NonFocus"/>.
+/// </summary>
+internal abstract partial class AsynchronousViewportTaggerProvider<TTag> : IViewTaggerProvider
+    where TTag : ITag
+{
+    private enum ViewPortToTag
+    {
+        InView,
+        AboveAndBelow,
+    }
```

Comment:
note: thsi could be a bool.  but i liked the enum for clarity.  Willing to change if people feel differently.